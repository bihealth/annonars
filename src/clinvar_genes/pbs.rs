//! Data structures for (de-)serialization as generated by `prost-build`.

include!(concat!(env!("OUT_DIR"), "/annonars.clinvar.v1.per_gene.rs"));
