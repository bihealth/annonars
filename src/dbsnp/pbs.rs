//! Data structures for (de-)serialization as generated by `prost-build`.

use noodles_vcf::variant::record::AlternateBases;

pub use crate::pbs::dbsnp::Record;
use noodles_vcf::variant::record_buf::info::field;

impl Record {
    /// Creates a new `Record` from a VCF record and allele number.
    pub fn from_vcf_allele(
        record: &noodles_vcf::variant::RecordBuf,
        allele_no: usize,
    ) -> Result<Self, anyhow::Error> {
        let chrom = record.reference_sequence_name().to_string();
        let pos: usize = record
            .variant_start()
            .expect("Telomeric breakends not supported")
            .get();
        let pos: i32 = i32::try_from(pos)?;
        let ref_allele = record.reference_bases().to_string();
        let alt_allele = record
            .alternate_bases()
            .iter()
            .nth(allele_no)
            .ok_or_else(|| anyhow::anyhow!("no such allele: {}", allele_no))??
            .to_string();
        let rs_id = if let Some(Some(field::Value::Integer(rs))) = record.info().get("RS") {
            *rs
        } else {
            anyhow::bail!("no rs id in dbSNP record")
        };

        Ok(Record {
            chrom,
            pos,
            ref_allele,
            alt_allele,
            rs_id,
        })
    }
}
