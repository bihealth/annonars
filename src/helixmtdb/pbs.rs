//! Data structures for (de-)serialization as generated by `prost-build`.

use std::str::FromStr;

pub use crate::pbs::helixmtdb::Record;
use noodles_vcf::record::info::field;

impl Record {
    /// Creates a new `Record` from a VCF record and allele number.
    pub fn from_vcf_allele(
        record: &noodles_vcf::record::Record,
        allele_no: usize,
    ) -> Result<Self, anyhow::Error> {
        let chrom = record.chromosome().to_string();
        let pos: usize = record.position().into();
        let pos = pos as i32;
        let ref_allele = record.reference_bases().to_string();
        let alt_allele = record
            .alternate_bases()
            .get(allele_no)
            .ok_or_else(|| anyhow::anyhow!("no such allele: {}", allele_no))?
            .to_string();
        let num_total = if let Some(Some(field::Value::Integer(num_total))) =
            record.info().get(&field::Key::from_str("AN")?)
        {
            *num_total
        } else {
            anyhow::bail!("missing INFO/AN in HelixMtDb record")
        };
        let num_het = if let Some(Some(field::Value::Integer(num_het))) =
            record.info().get(&field::Key::from_str("AC_het")?)
        {
            *num_het
        } else {
            anyhow::bail!("missing INFO/AC in HelixMtDb record")
        };
        let num_hom = if let Some(Some(field::Value::Integer(num_hom))) =
            record.info().get(&field::Key::from_str("AC_hom")?)
        {
            *num_hom
        } else {
            anyhow::bail!("missing INFO/AC_hom in HelixMtDb record")
        };
        let feature_type = if let Some(Some(field::Value::String(feature))) =
            record.info().get(&field::Key::from_str("FEATURE")?)
        {
            feature.to_string()
        } else {
            anyhow::bail!("missing INFO/FEATURE in HelixMtDb record")
        };
        let gene_name = if let Some(Some(field::Value::String(gene_name))) =
            record.info().get(&field::Key::from_str("GENE")?)
        {
            gene_name.to_string()
        } else {
            anyhow::bail!("missing INFO/GENE in HelixMtDb record")
        };

        Ok(Record {
            chrom,
            pos,
            ref_allele,
            alt_allele,
            num_total,
            num_het,
            num_hom,
            feature_type,
            gene_name,
        })
    }
}
