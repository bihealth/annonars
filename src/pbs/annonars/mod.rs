//! Code generate for protobufs by `prost-build`.

pub mod clinvar;
pub mod functional;
