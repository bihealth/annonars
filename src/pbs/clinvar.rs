//! Code generate for protobufs by `prost-build`.

/// Code generate for protobufs by `prost-build`.
pub mod minimal {
    include!(concat!(env!("OUT_DIR"), "/annonars.clinvar.minimal.rs"));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar.minimal.serde.rs"
    ));
}

/// Code generate for protobufs by `prost-build`.
pub mod per_gene {
    include!(concat!(env!("OUT_DIR"), "/annonars.clinvar.per_gene.rs"));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar.per_gene.serde.rs"
    ));
}

/// Code generate for protobufs by `prost-build`.
pub mod sv {
    include!(concat!(env!("OUT_DIR"), "/annonars.clinvar.sv.rs"));
    include!(concat!(env!("OUT_DIR"), "/annonars.clinvar.sv.serde.rs"));
}
