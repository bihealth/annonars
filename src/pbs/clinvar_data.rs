//! Code generate for protobufs by `prost-build`.

/// Code generated for protobufs by `prost-build`.
pub mod class_by_freq {
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.class_by_freq.rs"
    ));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.class_by_freq.serde.rs"
    ));
}

/// Code generated for protobufs by `prost-build`.
pub mod clinvar_public {
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.clinvar_public.rs"
    ));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.clinvar_public.serde.rs"
    ));

    impl Chromosome {
        /// Return the chromosome name, e.g., "1", ..., "22", "X", "Y", "MT", "PAR", "Un".
        pub fn as_chr_name(&self) -> String {
            self.as_str_name().replace("CHROMOSOME_", "")
        }
    }
}

/// Code generated for protobufs by `prost-build`.
pub mod extracted_vars {
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.extracted_vars.rs"
    ));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.extracted_vars.serde.rs"
    ));
}

/// Code generated for protobufs by `prost-build`.
pub mod gene_impact {
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.gene_impact.rs"
    ));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.gene_impact.serde.rs"
    ));
}

/// Code generated for protobufs by `prost-build`.
pub mod phenotype_link {
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.phenotype_link.rs"
    ));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.clinvar_data.phenotype_link.serde.rs"
    ));
}
