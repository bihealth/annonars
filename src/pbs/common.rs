//! Code generate for protobufs by `prost-build`.

/// Information about versions.
pub mod versions {
    include!(concat!(env!("OUT_DIR"), "/annonars.common.versions.rs"));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.common.versions.serde.rs"
    ));
}
