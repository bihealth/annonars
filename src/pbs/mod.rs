//! Code generate for protobufs by `prost-build`.

pub mod clinvar;
pub mod cons;
pub mod dbsnp;
pub mod functional;
pub mod genes;
pub mod gnomad;
pub mod helixmtdb;
pub mod panelapp;
pub mod regions;
