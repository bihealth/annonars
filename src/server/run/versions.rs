//! Provide information about the software and database versions.

use actix_web::{
    get,
    web::{self, Data, Json, Path},
    Responder,
};

use super::{error::CustomError, WebServerData};

/// Code for deserializing the version `spec.yaml` files.
pub mod schema {
    use std::path::Path;

    use crate::pbs;

    /// Information about input data.
    #[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
    pub struct CreatedFrom {
        /// Data source name.
        pub name: String,
        /// Version of the data source.
        pub version: String,
    }

    impl Into<pbs::common::versions::CreatedFrom> for CreatedFrom {
        fn into(self) -> pbs::common::versions::CreatedFrom {
            let Self { name, version } = self;
            pbs::common::versions::CreatedFrom { name, version }
        }
    }

    /// Version specification.
    #[serde_with::skip_serializing_none]
    #[serde_with::serde_as]
    #[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
    pub struct VersionSpec {
        /// Identifier of the data.
        #[serde(rename = "dc.identifier")]
        pub identifier: String,
        /// Title of the data.
        #[serde(rename = "dc.title")]
        pub title: String,
        /// Creator of the data.
        #[serde(rename = "dc.creator")]
        pub creator: String,
        /// Contributors of the data.
        #[serde(rename = "dc.contributor")]
        pub contributor: Option<Vec<String>>,
        /// Format of the data.
        #[serde(rename = "dc.format")]
        pub format: String,
        /// Date of the data.
        #[serde(rename = "dc.date")]
        pub date: String,
        /// Version of the data.
        #[serde(rename = "x-version")]
        pub version: String,
        /// Optional genome release.
        #[serde(rename = "x-genome-release")]
        pub genome_release: Option<String>,
        /// Data description.
        #[serde(rename = "dc.description")]
        pub description: String,
        /// Data source.
        #[serde(rename = "dc.source")]
        pub source: Vec<String>,
        /// Created from information.
        #[serde(rename = "x-created-from")]
        pub created_from: Vec<CreatedFrom>,
    }

    impl VersionSpec {
        /// Read a `VersionSpec` from a YAML file.
        pub fn from_path<P>(p: P) -> Result<Self, anyhow::Error>
        where
            P: AsRef<Path>,
        {
            let full_path = p.as_ref().to_str().ok_or_else(|| {
                anyhow::anyhow!("problem converting path to string: {:?}", p.as_ref())
            })?;
            let yaml_str = std::fs::read_to_string(&full_path)
                .map_err(|e| anyhow::anyhow!("problem reading file {}: {}", &full_path, e))?;
            serde_yaml::from_str(&yaml_str)
                .map_err(|e| anyhow::anyhow!("problem deserializing {}: {}", full_path, e))
        }
    }

    impl Into<pbs::common::versions::VersionSpec> for VersionSpec {
        fn into(self) -> pbs::common::versions::VersionSpec {
            let Self {
                identifier,
                title,
                creator,
                contributor,
                format,
                date,
                version,
                genome_release,
                description,
                source,
                created_from,
            } = self;
            pbs::common::versions::VersionSpec {
                identifier,
                title,
                creator,
                contributor: contributor.unwrap_or_default(),
                format,
                date,
                version,
                genome_release,
                description,
                source,
                created_from: created_from.into_iter().map(Into::into).collect(),
            }
        }
    }
}

/// Query parameters for `handle()`.
#[derive(Debug, Clone, serde::Deserialize)]
struct Request {
    pub genome_release: String,
}

/// Query for annotations for one variant.
#[get("/v1/versions")]
async fn handle(
    data: Data<WebServerData>,
    _path: Path<()>,
    query: web::Query<Request>,
) -> actix_web::Result<impl Responder, CustomError> {
    let genome_release =
        query
            .into_inner()
            .genome_release
            .parse()
            .map_err(|e: strum::ParseError| {
                CustomError::new(anyhow::anyhow!("problem getting genome release: {}", e))
            })?;
    Ok(Json(data.db_infos[genome_release].clone()))
}

#[cfg(test)]
pub mod test {
    #[rstest::rstest]
    #[case("alphamissense-grch37-1+0.33.0")]
    #[case("alphamissense-grch38-1+0.33.0")]
    #[case("functional-grch37-105.20201022+0.33.0")]
    #[case("functional-grch38-110+0.33.0")]
    #[case("genes-3.1+4.0+4.5+20230606+10.1+20240105+0.33.0")]
    #[case("gnomad-mtdna-grch37-3.1+0.33.0")]
    #[case("gnomad-mtdna-grch38-3.1+0.33.0")]
    #[case("helixmtdb-grch37-20200327+0.33.0")]
    #[case("helixmtdb-grch38-20200327+0.33.0")]
    fn test_deserialize_spec_yaml(#[case] name: &str) -> Result<(), anyhow::Error> {
        crate::common::set_snapshot_suffix!("{}", &name);

        let full_path = format!("tests/server/annonars/{}/spec.yaml", &name);
        let spec = super::schema::VersionSpec::from_path(&full_path)?;
        insta::assert_yaml_snapshot!(&spec);

        let proto_spec: crate::pbs::common::versions::VersionSpec = spec.into();
        insta::assert_yaml_snapshot!(&proto_spec);

        Ok(())
    }
}
