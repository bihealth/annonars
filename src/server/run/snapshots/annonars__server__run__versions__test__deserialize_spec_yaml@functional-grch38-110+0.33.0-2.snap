---
source: src/server/run/versions.rs
expression: "&proto_spec"
---
identifier: "annonars/functional:110+0.33.0-"
title: annonars functional elements RocksDB
creator: VarFish Developer Teams
format: application/x-rocksdb
date: "20240105"
version: 110+0.33.0
description: "RocksDB built from RefSeq Functional Elements (and other sources in\nthe future).\n"
source:
  - "PMID:34876495"
  - "https://www.ncbi.nlm.nih.gov/refseq/"
createdFrom:
  - name: RefSeq Functional Elements
    version: "110"
